plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DynamicFilteringJdbcSplitSource.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import io.trino.spi.connector.ConnectorSplitSource;
+import io.trino.spi.connector.DynamicFilter;
+import io.trino.spi.predicate.TupleDomain;
+
+import java.util.concurrent.CompletableFuture;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Attaches dynamic filter to {@link JdbcSplit} after {@link DynamicFilteringJdbcSplitManager}
+ * has waited for the collection of dynamic filters.
+ * This allows JDBC based connectors to avoid waiting for dynamic filters again on the worker node
+ * in {@link JdbcRecordSetProvider}. The number of splits generated by JDBC based connectors are
+ * typically small, therefore attaching dynamic filter here does not add significant overhead.
+ */
+public class DynamicFilteringJdbcSplitSource
+        implements ConnectorSplitSource
+{
+    private final ConnectorSplitSource delegateSplitSource;
+    private final DynamicFilter dynamicFilter;
+    private final JdbcTableHandle tableHandle;
+
+    DynamicFilteringJdbcSplitSource(ConnectorSplitSource delegateSplitSource, DynamicFilter dynamicFilter, JdbcTableHandle tableHandle)
+    {
+        this.delegateSplitSource = requireNonNull(delegateSplitSource, "delegateSplitSource is null");
+        this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+        this.tableHandle = requireNonNull(tableHandle, "tableHandle is null");
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+    {
+        return delegateSplitSource.getNextBatch(maxSize)
+                .thenApply(batch -> {
+                    TupleDomain<JdbcColumnHandle> dynamicFilterPredicate = dynamicFilter.getCurrentPredicate()
+                            .filter((columnHandle, domain) -> isColumnEligibleForDynamicFilter(tableHandle))
```

Comment:
this seems to be redundant to https://github.com/trinodb/trino/pull/13334/commits/2ce2dce546c6fe19fdaa77ae07428f48401cfc36#diff-0f107bc3ebdfca6a78610fcbf4047256cc9bdb749933e2a3b260e80c55c11586R73:
```
    public ConnectorSplitSource getSplits(
            ConnectorTransactionHandle transaction,
            ConnectorSession session,
            ConnectorTableHandle table,
            DynamicFilter dynamicFilter,
            Constraint constraint)
    {
        JdbcTableHandle tableHandle = (JdbcTableHandle) table;
        // pushing DF through limit could reduce query performance
        boolean hasLimit = tableHandle.getLimit().isPresent();
        if (dynamicFilter == DynamicFilter.EMPTY || hasLimit || !dynamicFilteringEnabled(session)) {
            return delegateSplitManager.getSplits(transaction, session, table, dynamicFilter, constraint);
        }
```

Reply:
Bindings for ConnectorSplitManager in JDBC module are extensible, so it's safer to not assume that this check was already applied